embedded-nal-async/src/stack/udp.rs

```diff
@@ -0,0 +1,232 @@
+	where
+		Self: 'a;
+
+	/// Receive a datagram into the provided buffer.
+	///
+	/// If the received datagram exceeds the buffer's length, it is received regardless, and the
+	/// remaining bytes are discarded. The full datagram size is still indicated in the result,
+	/// allowing the recipient to detect that truncation.
+	///
+	/// The local and remote address are given, in that order, in the result along with the number
+	/// of bytes.
+	fn receive_into<'a>(&'a mut self, buffer: &'a mut [u8]) -> Self::ReceiveIntoFuture<'a>;
+	/// Return type of the [`.receive_into()`] method
+	type ReceiveIntoFuture<'a>: Future<
+		Output = Result<(usize, SocketAddr, SocketAddr), Self::Error>,
+	> where
+		Self: 'a;
+}
+
+/// This trait is implemented by UDP/IP stacks. The trait allows the underlying driver to
+/// construct multiple connections that implement the I/O traits from embedded-io.
+///
+/// Note that stacks with exotic connection creation methods may still not implement this, yet have
+/// objects that implement [`ConnectedUdp`] or similar.
+pub trait UdpStack {
+	/// Error type returned on socket creation failure.
+	type Error: embedded_io::Error;
+
+	/// Eventual socket return type of the [`.connect()`] method
+	type Connected<'m>: ConnectedUdp
+	where
+		Self: 'm;
+	/// Eventual socket return type of the [`.bind_single()`] method
+	type UniquelyBound<'m>: UnconnectedUdp
+	where
+		Self: 'm;
+	/// Eventual return type of the [`.bind_multiple()`] method
+	type MultiplyBound<'m>: UnconnectedUdp
+	where
+		Self: 'm;
+
+	/// Create a socket that has a fixed remote address.
+	///
+	/// The local address is chosen automatically.
+	///
+	/// While asynchronous traits implemented through GAT can not have provided default methods,
+	/// implementers are encouraged to use the hidden `.connect_default()` method if all they would
+	/// do is delegating to [`.connect_from`] with a suitable unspecified local address.
+	fn connect(&self, remote: SocketAddr) -> Self::ConnectFuture<'_>;
+	/// Future return type of the [`.connect()`] method
+	type ConnectFuture<'a>: Future<Output = Result<(SocketAddr, Self::Connected<'a>), Self::Error>>
+	where
+		Self: 'a;
+
+	/// Create a socket that has a fixed remote address.
+	///
+	/// The local address is given explicitly, but may be partially unspecified; it is fixed by the
+	/// network stack at connection time. The full local address is returned along with the
+	/// connected socket, primarily for debugging purposes.
+	fn connect_from(&self, local: SocketAddr, remote: SocketAddr) -> Self::ConnectFromFuture<'_>;
+	/// Future return type of the [`.connect_from()`] method
+	type ConnectFromFuture<'a>: Future<
+		Output = Result<(SocketAddr, Self::Connected<'a>), Self::Error>,
+	> where
+		Self: 'a;
+
+	/// Helper that implements [`connect()`] using [`connect_from()`].
+	#[doc(hidden)]
+	fn connect_default(&self, remote: SocketAddr) -> Self::ConnectFromFuture<'_> {
```

Comment:
I don't understand how this differs from connect() in practice. If the local address is unspecified, isn't the only reasonable thing for the implementation to select one automatically like connect()? Unless it's a common thing to do, I would probably leave it out of the trait.

Reply:
As (for lack of async fn in traits) the connect() method is paired with the ConnectFuture type, we can't have provided methods. Were that not the case, this could be the default impl for `connect()`, but as things are, implementers who want to use the simple default can just set `type ConnectFuture<'a> = ConnectFromFuture<'a>` and `fn connect(&self, remote) -> ... { self.connect_default(remote) }` in lieu of using the provided method.

It is a distinct method with its distinct return type because calling `.connect()` will be a common occurrence, and implementers might want to optimize this over the provided default. (In a POSIX implementation, this might be saving a syscall).

Reply:
This is not documented here (because the item is hidden, because users should not rely on it), but in the documentation of the `.connect()` method.